base/iterators.jl

```diff
@@ -1032,7 +1032,7 @@ _prod_eltype(::Type{I}) where {I<:Tuple} =
 iterate(::ProductIterator{Tuple{}}) = (), true
 iterate(::ProductIterator{Tuple{}}, state) = nothing
 
-@inline isdone(P::ProductIterator) = any(isdone, P.iterators)
+@inline isdone(P::ProductIterator) = any(Base.map(isdone, P.iterators))  # issue #40283
```

Comment:
We should make sure to keep the sort-circuiting behavior here and not unconditionally evaluate `isdone` for every entry in `P.iterators`. I agree with @N5N3 that there should probably be an optimized version of `any` for this.

Reply:
Got it, in that case we can add a version of `any` for tuples. I guess the same should be done for `all`, which seems to have the same issue.